subxt/src/storage/storage_key.rs

```diff
@@ -0,0 +1,232 @@
+use super::{
+    storage_address::StaticStorageKey,
+    utils::{strip_concat_hash_bytes, strip_storage_addess_root_bytes},
+};
+use crate::{
+    dynamic::DecodedValueThunk,
+    error::{Error, StorageAddressError},
+    metadata::{DecodeWithMetadata, Metadata},
+    utils::{Encoded, Static},
+};
+use scale_encode::EncodeAsType;
+use subxt_metadata::StorageHasher;
+
+/// This trait should be implemented by anything that can be used as one or multiple storage keys.
+pub trait StorageKey {
+    /// Iterator over the storage keys, each key implements EncodeAsType to
+    /// give the corresponding bytes to a `StorageHasher`.
+    fn keys_iter(&self) -> impl ExactSizeIterator<Item = &dyn EncodeAsType>;
+
+    /// Attempts to decode the StorageKey from a whole storage address.
+    /// The key/keys can only be recovered if all hashers are concat-style hashers.
+    /// Example: Imagine The `StorageKey` is a tuple (A,B) and the hashers are: [Blake2_128Concat, Twox64Concat].
+    /// Then the memory layout of the storage key is:
+    /// ```txt
+    /// | 8 bytes pallet hash | 8 bytes entry hash | 16 bytes hash of A | ... bytes of A | 8 bytes hash of B | ... bytes of B |
+    /// ```
+    /// Returns None, if any of the hashers is not a concat-style hasher.
+    fn decode_from_address_bytes(
+        address_bytes: &[u8],
+        hashers_and_ty_ids: &[(StorageHasher, u32)],
+        metadata: &Metadata,
+    ) -> Option<Result<Self, Error>>
+    where
+        Self: Sized + 'static;
+}
+
+/// Implement `StorageKey` for `()` which can be used for keyless storage entries
+impl StorageKey for () {
+    fn keys_iter(&self) -> impl ExactSizeIterator<Item = &dyn EncodeAsType> {
+        // Note: this returns the storage root address of the storage entry.
+        // It gives the same result as if you were to use `vec![]` as a `StorageKey`.
+        std::iter::empty()
+    }
+
+    fn decode_from_address_bytes(
+        address_bytes: &[u8],
+        _hashers_and_ty_ids: &[(StorageHasher, u32)],
+        _metadata: &Metadata,
+    ) -> Option<Result<Self, Error>> {
+        // We need at least 16 bytes, becauase there are 8 bytes for the pallet hash and
+        // another 8 bytes for the entry hash at the beginning of each storage address.
+        if address_bytes.len() < 16 {
+            return Some(Err(StorageAddressError::UnexpectedAddressBytes.into()));
+        }
```

Comment:
I wonder whether the bytes given to a `StorageKey` should only ever be for the subset of things that this `StorageKey` impl is expected to handle?

Then, a more generic tuple impl for `StorageKey` would do nothing except look at the hashers and delegate the relevant bytes and hasher impl to each entry in the tuple (or error if mismatch in count of entries).

Each impl would also then have it easy and doesn't need to trim any bytes from the start; it would be given exactly the bytes and hashers that it is expected to work with.

Then the general storage code that calls into this trait would just trim the relevant root bytes from the front once.